c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,522 @@
+                slf.emit_str("]");
+            }),
+            Rewrite::SliceTail(ref arr, ref idx) => self.emit_parenthesized(prec > 3, |slf| {
+                slf.emit_rewrite(arr, 3, emit_expr, emit_subexpr);
+                slf.emit_str("[");
+                // Rather than figure out the right precedence for `..`, just force
+                // parenthesization in this position.
+                slf.emit_rewrite(idx, 999, emit_expr, emit_subexpr);
+                slf.emit_str(" ..]");
+            }),
+            Rewrite::CastUsize(ref rw) => self.emit_parenthesized(prec > 1, |slf| {
+                slf.emit_rewrite(rw, 1, emit_expr, emit_subexpr);
+                slf.emit_str(" as usize");
+            }),
+            Rewrite::LitZero => {
+                self.emit_str("0");
+            }
+
+            Rewrite::PrintTy(ref s) => {
+                self.emit_str(s);
+            }
+            Rewrite::TyPtr(ref rw, mutbl) => {
+                match mutbl {
+                    Mutability::Not => self.emit_str("*const "),
+                    Mutability::Mut => self.emit_str("*mut "),
+                }
+                self.emit_rewrite(rw, 0, emit_expr, emit_subexpr);
+            }
+            Rewrite::TyRef(ref rw, mutbl) => {
+                match mutbl {
+                    Mutability::Not => self.emit_str("&"),
+                    Mutability::Mut => self.emit_str("&mut "),
+                }
+                self.emit_rewrite(rw, 0, emit_expr, emit_subexpr);
+            }
+            Rewrite::TySlice(ref rw) => {
+                self.emit_str("[");
+                self.emit_rewrite(rw, 0, emit_expr, emit_subexpr);
+                self.emit_str("]");
+            }
+            Rewrite::TyCtor(ref name, ref rws) => {
+                self.emit_str(name);
+                self.emit_str("<");
+                for rw in rws {
+                    self.emit_rewrite(rw, 0, emit_expr, emit_subexpr);
+                }
+                self.emit_str(">");
+            }
+        }
+    }
+
+    fn emit_bytes(&mut self, lo: BytePos, hi: BytePos) {
+        assert!(
+            self.file.start_pos <= lo && hi <= self.file.end_pos,
+            "bytes {:?} .. {:?} are out of range for file {:?}",
+            lo,
+            hi,
+            self.file.name
+        );
+        let src = self
+            .file
+            .src
+            .as_ref()
+            .unwrap_or_else(|| panic!("source is not available for file {:?}", self.file.name));
+        self.emit_str(&src[lo.0 as usize..hi.0 as usize]);
+    }
+
+    fn emit_span_with_rewrites(&mut self, span: Span, rts: &[RewriteTree]) {
+        let (_, overlap, _) = partition_nodes(rts, span);
+
+        let mut pos = span.lo();
+        for rt in overlap {
+            // Every child node is contained by the span of some `Rewrite::Identity` or
+            // `Rewrite::Sub` in its parent node.
+            debug_assert!(span.contains(rt.span));
```

Comment:
would this assert be more appropriately placed in the partition function?

Reply:
This is slightly stricter than the guarantee provided by `partition_nodes`, which implies only that `span.overlaps(rt.span)`.